File: src/test/java/com/libraryjava/service/BookServiceTest.java
package com.libraryjava.service;

import com.libraryjava.domain.Book;
import com.libraryjava.domain.UserLoanHistory;
import com.libraryjava.domain.user.User;
import com.libraryjava.domain.user.UserStatus;
import com.libraryjava.dto.book.BookLoanDto;
import com.libraryjava.dto.book.BookMakeDto;
import com.libraryjava.dto.book.BookReturnDto;
import com.libraryjava.repository.BookRepository;
import com.libraryjava.repository.UserLoanHistoryRepository;
import com.libraryjava.repository.UserRepository;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

import static org.assertj.core.api.Assertions.*;

@SpringBootTest
class BookServiceTest {
    @Autowired
    BookService bookService;
    @Autowired
    BookRepository bookRepository;
    @Autowired
    UserRepository userRepository;
    @Autowired
    UserLoanHistoryRepository userLoanHistoryRepository;


    @AfterEach
    void clear() {
        bookRepository.deleteAll();
        userRepository.deleteAll();
    }

    @Test
    @DisplayName("책 등록이 정상 동작한다.")
    void makeBookTest() {
        // given
        BookMakeDto request = new BookMakeDto("이상한 나라의 앨리스", "동화");

        // when
        bookService.makeBook(request);

        // then
        List<Book> books = bookRepository.findAll();
        assertThat(books).hasSize(1);
    }

    @Test
    @DisplayName("책 대출 성공 테스트")
    void loanBookSuccessTest() {
        // given
        bookRepository.save(Book.fixture("이상한 나라의 앨리스", "소설"));
        User savedUser = userRepository.save(User.fixture("김진규", 10));
        BookLoanDto request = new BookLoanDto("김진규", "이상한 나라의 앨리스");

        // when
        bookService.loanBook(request);

        // then
        List<UserLoanHistory> results = userLoanHistoryRepository.findAll();
        assertThat(results).hasSize(1);
        System.out.println("results.get(0) = " + results.get(0));
        assertThat(results.get(0).getBookName()).isEqualTo("이상한 나라의 앨리스");
        System.out.println("results.get(0) = " + results.get(0));
        assertThat(results.get(0).getUser().getId()).isEqualTo(savedUser.getId());
        System.out.println("results.get(0) = " + results.get(0));
        assertThat(results.get(0).getUserStatus()).isEqualTo(UserStatus.ACTIVE);
    }

    @Test
    @DisplayName("책 대출 실패 테스트")
    void loanBookFailTest() {
        // given
        bookRepository.save(Book.fixture("이상한 나라의 앨리스", "소설"));
        User savedUser = userRepository.save(User.fixture("김진규", 10));
        userLoanHistoryRepository.save(new UserLoanHistory("이상한 나라의 앨리스", UserStatus.ACTIVE, savedUser));
        BookLoanDto request = new BookLoanDto("김진규", "이상한 나라의 앨리스");

        // when & then
        Assertions.assertThrows(IllegalArgumentException.class, () -> bookService.loanBook(request), "진작 대출되어 있는 책입니다.");
    }

    @Test
    @DisplayName("책 반납이 정상 동작한다.")
    void returnBookTest() {
        // given
        bookRepository.save(Book.fixture("이상한 나라의 앨리스", "소설"));
        User savedUser = userRepository.save(User.fixture("김진규", 10));
        userLoanHistoryRepository.save(new UserLoanHistory("이상한 나라의 앨리스", UserStatus.ACTIVE, savedUser));
        BookReturnDto request = new BookReturnDto("김진규", "이상한 나라의 앨리스");

        // when
        bookService.returnBook(request);

        // then
        List<UserLoanHistory> results = userLoanHistoryRepository.findAll();
        assertThat(results).hasSize(1);
        assertThat(results.get(0).getUserStatus()).isEqualTo(UserStatus.IN_ACTIVE);
    }
}

File: src/test/kotlin/com/library/service/book/BookServiceTest.kt
package com.library.service.book

import com.library.domain.book.Book
import com.library.domain.book.BookRepository
import com.library.domain.user.User
import com.library.domain.user.UserRepository
import com.library.domain.user.loanhistory.UserLoanHistoryRepository
import com.library.dto.book.BookLoanRequest
import com.library.dto.book.BookRequest
import com.library.dto.book.BookReturnRequest
import org.assertj.core.api.AssertionsForInterfaceTypes.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import java.lang.IllegalArgumentException

@SpringBootTest
class BookServiceTest @Autowired constructor(
    private val bookService: BookService,
    private val bookRepository: BookRepository,
    private val userRepository: UserRepository,
    private val loanHistoryRepository: UserLoanHistoryRepository,
) {
    @AfterEach
    fun clear() {
        bookRepository.deleteAll()
        userRepository.deleteAll()
        loanHistoryRepository.deleteAll()
    }

    @Test
    @DisplayName("책 저장 테스트")
    fun saveBookTest() {
        // given
        val request = BookRequest("책1")

        // when
        bookService.saveBook(request)

        // then
        val books = bookRepository.findAll()
        assertThat(books).hasSize(1)
    }

    @Test
    @DisplayName("책 대출 성공 테스트")
    fun loanBookSuccessTest() {
        // given
        bookRepository.save(Book("책1"))
        userRepository.save(User("유저1", null))
        val bookLoanRequest = BookLoanRequest("책1", "유저1")

        // when
        bookService.loanBook(bookLoanRequest)

        // then
        val loanHistories = loanHistoryRepository.findAll()
        assertThat(loanHistories).hasSize(1)
        assertThat(loanHistories[0].bookName).isEqualTo("책1")
        assertThat(loanHistories[0].isReturn).isFalse()
        assertThat(loanHistories[0].user.name).isEqualTo("유저1")
    }

    @Test
    @DisplayName("책 대출 실패 테스트")
    fun loanBookFailTest() {
        // given
        bookRepository.save(Book("책1"))
        userRepository.save(User("유저1", null))
        userRepository.save(User("유저2", null))
        val bookLoanRequest1 = BookLoanRequest("책1", "유저1")
        val bookLoanRequest2 = BookLoanRequest("책1", "유저2")

        // when
        bookService.loanBook(bookLoanRequest1)

        // then
        assertThrows<IllegalArgumentException> {
            bookService.loanBook(bookLoanRequest2)
        }.apply {
            assertThat(message).isEqualTo("진작 대출되어 있는 책입니다.")
        }
    }

    @Test
    @DisplayName("책 대출 반납 테스트")
    fun returnBookTest() {
        // given
        bookRepository.save(Book("책1"))
        userRepository.save(User("유저1", null))
        val bookLoanRequest = BookLoanRequest("책1", "유저1")
        bookService.loanBook(bookLoanRequest)
        val bookReturnRequest = BookReturnRequest("유저1", "책1")

        // when
        bookService.returnBook(bookReturnRequest)

        // then
        val loanHistories = loanHistoryRepository.findAll()
        assertThat(loanHistories).hasSize(1)
        assertThat(loanHistories[0].isReturn).isTrue()
    }
}